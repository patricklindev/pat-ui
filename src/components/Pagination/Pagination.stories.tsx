import React from 'react';
import Pagination from './Pagination';



export default {
  title: 'Pagination',
  component: Pagination,
};

export const Filled = () => 
  (
    <div>
      <Pagination
        totalPages={5}
        pgnType={'filled'}
        pgnShape={'basic'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'filled'}
        pgnShape={'basic'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'filled'}
        pgnShape={'rounded'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'filled'}
        pgnShape={'rounded'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
    </div>
    );

export const Outlined = () => 
  (
    <div>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'basic'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'basic'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'rounded'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'rounded'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
    </div>
    );

