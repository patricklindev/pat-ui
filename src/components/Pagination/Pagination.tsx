import React, { FC, useState } from 'react';
import Dropdown from 'react-dropdown';
import 'react-dropdown/style.css';

export type ButtonType = 'primary' | 'secondary' | 'default';

export interface PaginationProps {
    defaultpage: number;
    count: number;
    rowsPerPage?: number;
    TablePagination?: boolean;
    color?: ButtonType;
    disabled?: boolean;
    onChange?: Function;
}

/**
 * A Pagination enables the user to select a specific page from a range of pages.
 *
 * ```js
 * import {Pagination} from 'pat-ui'
 * ```
 */
export const PaginationContainer: FC<PaginationProps> = (props) => {
    const {
        rowsPerPage,
        defaultpage,
        count,
        TablePagination,
        color,
        disabled,
        onChange
    } = props;
    const [page, setPage] = useState(defaultpage);
    const [curRowsPerPage, setCurRowsPerPage] = useState(rowsPerPage);
    let options: string[] = [];
    if (rowsPerPage && TablePagination) {
        options = [
            rowsPerPage.toString(),
            Math.ceil(count / 3).toString(),
            Math.ceil(count / 2).toString(),
            count.toString(),
        ];
    }

    const handleChange = (newPage: number) => {
        if (onChange !== undefined) {
            onChange(newPage)
        }
        setPage(newPage);
    };

    return (
        <div>
            {TablePagination && rowsPerPage && curRowsPerPage ? (
                <div className="tablepagination">
                    <div>Rows per page: </div>
                    <Dropdown
                        className="tablepagination__dropdown"
                        options={options}
                        onChange={(e) => setCurRowsPerPage(+e.value)}
                        value={rowsPerPage.toString()}
                        placeholder={curRowsPerPage.toString()}
                    />
                    <div>
                        {page % curRowsPerPage === 0
                            ? (Math.floor(page / curRowsPerPage) - 1) * curRowsPerPage + 1
                            : Math.floor(page / curRowsPerPage) * curRowsPerPage + 1}
                        -
                        {page % curRowsPerPage === 0
                            ? Math.floor(page / curRowsPerPage) * curRowsPerPage
                            : Math.floor(page / curRowsPerPage) * curRowsPerPage +
                            curRowsPerPage}{' '}
                        of {count}
                    </div>
                    <button
                        className={
                            page - curRowsPerPage < 1
                                ? 'tablepagination__btn-disabled'
                                : 'tablepagination__btn'
                        }
                        disabled={page - curRowsPerPage < 1}
                        onClick={() => handleChange(page - curRowsPerPage)}
                    >
                        &lt;
                    </button>
                    <button
                        className={
                            page + curRowsPerPage > count
                                ? 'tablepagination__btn-disabled'
                                : 'tablepagination__btn'
                        }
                        disabled={page + curRowsPerPage > count}
                        onClick={() => handleChange(page + curRowsPerPage)}
                    >
                        &gt;
                    </button>
                </div>
            ) : (
                <div className="pagination">
                    <button
                        className={
                            page === 1 ? 'pagination__btn--disabled' : disabled ? 'pagination__btn--disabled' : 'pagination__btn'
                        }
                        disabled={page === 1 || disabled}
                        onClick={() => handleChange(page - 1)}
                        type="button"
                        data-testid="lt"
                    >
                        &lt;
                    </button>
                    <button
                        className={page === 1 ? color : disabled ? 'pagination__btn--disabled' : 'pagination__btn'}
                        disabled={disabled}
                        onClick={() => handleChange(1)}
                        type="button"
                    >
                        {1}
                    </button>
                    {page > 3 && <div style={{ color: 'grey' }}>...</div>}

                    {
                        page > 2 && (
                            <button
                                className={disabled ? 'pagination__btn--disabled' : "pagination__btn"}
                                disabled={disabled}
                                onClick={() => handleChange(page - 1)}
                                type="button"
                            >
                                {page - 1}
                            </button>
                        )
                    }

                    {
                        page !== 1 && page !== count && (
                            <button
                                className={`${disabled ? 'pagination__btn--disabled--selected' : color}`}
                                disabled={disabled}
                                onClick={() => handleChange(page)}
                                type="button"
                            >
                                {page}
                            </button>
                        )
                    }

                    {
                        page < count - 1 && (
                            <button
                                className={disabled ? 'pagination__btn--disabled' : "pagination__btn"}
                                disabled={disabled}
                                onClick={() => handleChange(page + 1)}
                                type="button"
                            >
                                {page + 1}
                            </button>
                        )
                    }

                    {page < count - 2 && <div style={{ color: 'grey' }}>...</div>}

                    <button
                        className={page === count ? color : disabled ? 'pagination__btn--disabled' : 'pagination__btn'}
                        disabled={disabled}
                        onClick={() => handleChange(count)}
                        type="button"
                    >
                        {count}
                    </button>

                    <button
                        className={
                            page === count ? 'pagination__btn--disabled' : disabled ? 'pagination__btn--disabled' : 'pagination__btn'
                        }
                        disabled={page === count || disabled}
                        onClick={() => handleChange(page + 1)}
                        type="button"
                        data-testid="gt"
                    >
                        &gt;
                    </button>
                </div >
            )}
        </div >
    );
};

PaginationContainer.defaultProps = {
    defaultpage: 1,
    count: 20,
    rowsPerPage: 5,
    TablePagination: false,
    color: 'default',
    disabled: false,
};

export default PaginationContainer;
