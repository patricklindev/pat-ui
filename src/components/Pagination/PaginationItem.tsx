import React,{ FC } from 'react'
type liTypes = React.DetailedHTMLProps<React.HTMLAttributes<HTMLLIElement>, HTMLLIElement>
interface IPaginationItem { 
    callBack: ()=> void;
    builtClass?: string
} 

const PaginationItem: FC<IPaginationItem> = (props)=>{
    const { builtClass ='', callBack, children, ...rest} = props
    
    function onClickHandler(e:React.MouseEvent<HTMLDivElement>): void{ 
        callBack()
    }
    
    return(
        <div data-testid='pageButtonOuterHolder' role={'button'} className={builtClass} onClick={onClickHandler}>
        <li data-testid='pageButtons' {...(rest as liTypes)} >
           {children}
        </li>
        </div>
    )
}

export default PaginationItem
