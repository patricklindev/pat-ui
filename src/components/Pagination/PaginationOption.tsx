import React, {
  ButtonHTMLAttributes,
  AnchorHTMLAttributes,
  FC,
  MouseEvent,
} from 'react';
import { classNames } from '../../utils/classNames';

export type ButtonSize = 'lg' | 'sm';
export type ButtonType =
  | 'primary'
  | 'secondary'
  | 'danger'
  | 'default'
  | 'link';

export interface IButtonProps {
  /** set customized style */
  className?: string;
  /** set button size */
  btnSize?: ButtonSize;
  /** set button type */
  btnType?: ButtonType;
  /** set disabled button */
  disabled?: boolean;
  selected?: boolean;
  shape: string;
  variant?: string;
}

type NativeButtonProps = IButtonProps & ButtonHTMLAttributes<HTMLButtonElement>;
type NativeAchorButtonProps = IButtonProps &
  AnchorHTMLAttributes<HTMLAnchorElement>;
export type PatButtonProps = NativeButtonProps | NativeAchorButtonProps;

export const PaginationOption: FC<PatButtonProps> = (props) => {
  const {
    btnSize,
    btnType,
    children,
    disabled,
    selected,
    className,
    shape,
    variant,
    ...rest
  } = props;
  let styleClasses = classNames('pagination--btn', {
    [`pagination--btn-${btnType}`]: true,
    [`pagination--btn-${btnSize}`]: !!btnSize,
    disabled: !!(disabled && btnType === 'link'),
    selected: !!selected,
    round: !!(shape === 'round'),
    outlined: !!(variant === 'outlined'),
  });

  if (className) {
    styleClasses += ' ' + className;
  }

  let btn;
  if (btnType !== 'link') {
    btn = (
      <button
        className={styleClasses}
        disabled={disabled}
        {...(rest as NativeButtonProps)}
      >
        {props.children}
      </button>
    );
  } else {
    if (disabled) {
      rest.onClick = (e: MouseEvent) => {
        e.preventDefault();
      };
    }
    btn = (
      <a
        className={styleClasses}
        disabled={disabled}
        selected={selected}
        {...(rest as NativeAchorButtonProps)}
      >
        {props.children}
      </a>
    );
  }
  return btn;
};

PaginationOption.defaultProps = {
  btnType: 'default',
  disabled: false,
};

export default PaginationOption;
