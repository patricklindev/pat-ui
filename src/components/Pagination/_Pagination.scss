.pgn-outlined.pgn-primary {
    .btn[id] {
        background-color: #ffd8e6;
        border: 1px solid #F50057;
        color: #F50057;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: black;
        }
    }
  }


  .pgn-outlined.pgn-secondary {
    .btn[id] {
        background-color: #dde2ff;
        border: 1px solid #3F51B5;
        color: #3F51B5;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
  }


  .pgn-filled.pgn-primary{
    .btn[id] {
        border: none;
        background-color: #F50057;
        color: white;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }

}

.pgn-filled.pgn-secondary{
    .btn[id] {
        border: none;
        background-color: #3F51B5;
        color: white;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
}


.pgn-basic{
    display: flex;
    flex-direction: row;
        .btn {
            border-radius: 100%;
            margin: 2px;
            display: flex;
            align-items: center;
            justify-content: center;
        }

}

.pgn-rounded{
    display: flex;
    flex-direction: row;
        .btn {
            border-radius: 10%;
            margin: 2px;
            align-items: center;
            display: flex;
            justify-content: center;
        }

}



.pgn-md {
    .btn{
        width: 32px;
        height: 32px;
    }
    .elip {
        display: flex;
        align-self: flex-end;
        font-size: 1.5rem;
    }
    .pgn-prev-next-btn{
        width: 32px;
        height: 32px;
        border: none;
    }
}


.pgn-lg {
    .btn{
        width: 40px;
        height: 40px;
    }
    .elip {
        display: flex;
        align-self: flex-end;
        font-size: 2rem;
    }
    .pgn-prev-next-btn{
        width: 40px;
        height: 40px;
        border: none;
    }
}

.pgn-sm {
    .btn{
        width: 26px;
        height: 26px;
    }
    .elip {
        display: flex;
        align-self: flex-end;
        font-size: 1rem;
    }
    .pgn-prev-next-btn{
        width: 26px;
        height: 26px;
        border: none;
    }
}
