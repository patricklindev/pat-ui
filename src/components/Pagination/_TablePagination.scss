.table-pagination {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: $list-inline-padding * 3;

  .table-pagination__row__option {
    display: flex;

    .table-pagination__select {
      border: none;

      &:hover {
        cursor: pointer;
      }
    }
  }

  .pagination__icons__btn-icon {
    display: flex;
    justify-content: center;
    align-items: center;
    background: $white;
    border: none;
    outline: none;
    border-radius: $border-radius-sm;
  }

  .pagination__icons--disabled {
    fill: $gray-400;
  }

  .pagination__icons--default {
    &.pagination__icons__btn-icon {
      @include button-size(
        $btn-padding-y-default,
        $btn-padding-x-default,
        $btn-font-size-default,
        $border-radius-round
      );
    }
    &.pagination__icons__btn-arrow {
      width: $btn-padding-x-default * 4;
      height: $btn-padding-x-default * 4;
    }
    &:hover:not(.pagination__icons--disabled) {
      background-color: $gray-200;
    }
  }
}
