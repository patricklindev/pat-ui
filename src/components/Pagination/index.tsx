export { default as Pagination } from './Pagination';
export { default as TablePagination } from './TablePaginationContainer';
