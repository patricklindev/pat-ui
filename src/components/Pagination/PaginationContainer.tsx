//this component is for story book and test only
import React, {
  useState,
  ButtonHTMLAttributes,
  AnchorHTMLAttributes,
  FC,
} from 'react';

import Pagination from './Pagination';

export type PaginationSize = 'lg' | 'sm';
export type PaginationType =
  | 'primary'
  | 'secondary'
  | 'danger'
  | 'default'
  | 'link';

export interface IPaginationProps {
  /** set customized style */
  className?: string;
  /** set button size */
  pagSize?: PaginationSize;
  /** set button type */
  pagType?: PaginationType;
  /** set disabled button */
  disabled?: boolean;
  /** set pagination number */
  totalPageNumber?: number;
  currentPage?: number;
  rowsPerPage?: number;
  shape?: string;
  variant?: string;
}

type NativeButtonProps = IPaginationProps &
  ButtonHTMLAttributes<HTMLButtonElement>;
type NativeAchorButtonProps = IPaginationProps &
  AnchorHTMLAttributes<HTMLAnchorElement>;
export type PaginationProps = NativeButtonProps | NativeAchorButtonProps;

export const PaginationContainer: FC<PaginationProps> = ({
  pagType = 'default',
  pagSize,
  shape = 'rounded',
  disabled = false,
  variant = '',
}) => {
  const [page, setPage] = useState(1);

  const handleClickPage = (p: number) => {
    setPage(p);
  };

  return (
    <Pagination
      currentPage={page}
      onPageChange={handleClickPage}
      pagType={pagType}
      pagSize={pagSize}
      shape={shape}
      disabled={disabled}
      variant={variant}
    >
      Default Pagination
    </Pagination>
  );
};

export default PaginationContainer;
