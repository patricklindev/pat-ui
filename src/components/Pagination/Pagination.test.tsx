import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';
import Pagination, { PaginationProps } from './Pagination';

describe('Pagination', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(<Pagination defaultpage={1} count={10}></Pagination>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should see first page, last page, and two adjacent pages and hide other pages with ellipsis', () => {
        const wrapper = render(<Pagination defaultpage={16} count={30}></Pagination>);
        const adj1 = wrapper.queryByText('15') as HTMLElement;
        const adj2 = wrapper.queryByText('16') as HTMLElement;
        const firstpage = wrapper.queryByText('1') as HTMLElement;
        const lastpage = wrapper.queryByText('30') as HTMLElement;
        const ellipsis = wrapper.queryAllByText('...');
        expect(adj1).toBeInTheDocument();
        expect(adj2).toBeInTheDocument();
        expect(firstpage).toBeInTheDocument();
        expect(lastpage).toBeInTheDocument();
        expect(ellipsis.length).toBe(2);
    })

    it('should use left and right button to switch pages', () => {
        const wrapper = render(<Pagination defaultpage={16} count={30}></Pagination>);
        const left = screen.getByTestId('lt') as HTMLElement;
        const right = screen.getByTestId('gt') as HTMLElement;
        expect(left).toBeInTheDocument();
        expect(right).toBeInTheDocument();
        fireEvent.click(left)
        const prev = wrapper.queryByText('14')
        expect(prev).toBeInTheDocument();
        fireEvent.click(right)
        const post = wrapper.queryByText('17')
        expect(right).toBeInTheDocument();
    })

    it('should be able to pass total pages, current pages, row pages, and table pagination as props', () => {
        const props: PaginationProps = {
            rowsPerPage: 5, defaultpage: 16, count: 30, TablePagination: true
        };
        const wrapper = render(<Pagination {...props}></Pagination>);
        const total = wrapper.queryByText('16-20 of 30')
        const dropdown_bar = wrapper.getByText('5')
        expect(dropdown_bar).toBeInTheDocument();
        expect(dropdown_bar.parentElement?.parentElement?.className).toBe("Dropdown-root tablepagination__dropdown")
        expect(total).toBeInTheDocument();
    })

})