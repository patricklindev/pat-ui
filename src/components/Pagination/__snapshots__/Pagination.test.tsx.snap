// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="pagination"
    >
      <button
        class="pagination__btn--disabled"
        data-testid="lt"
        disabled=""
        type="button"
      >
        &lt;
      </button>
      <button
        class="default"
        type="button"
      >
        1
      </button>
      <button
        class="pagination__btn"
        type="button"
      >
        2
      </button>
      <div
        style="color: grey;"
      >
        ...
      </div>
      <button
        class="pagination__btn"
        type="button"
      >
        10
      </button>
      <button
        class="pagination__btn"
        data-testid="gt"
        type="button"
      >
        &gt;
      </button>
    </div>
  </div>
</DocumentFragment>
`;
